pkg/controller/ps/upgrade.go

```diff
@@ -0,0 +1,149 @@
+package ps
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	apiv1alpha1 "github.com/percona/percona-server-mysql-operator/api/v1alpha1"
+)
+
+const controllerRevisionHash = "controller-revision-hash"
+
+func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *appsv1.StatefulSet, cr *apiv1alpha1.PerconaServerMySQL) error {
+	log := logf.FromContext(ctx)
+
+	if cr.Spec.Pause {
+		return nil
+	}
+
+	currentSet := sts
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Name:      sts.Name,
+		Namespace: sts.Namespace,
+	}, currentSet)
+	if err != nil {
+		return errors.Wrap(err, "failed to get current sfs")
+	}
+
+	pods := corev1.PodList{}
+	if err := r.Client.List(ctx, &pods, &client.ListOptions{
+		Namespace:     currentSet.Namespace,
+		LabelSelector: labels.SelectorFromSet(currentSet.Labels),
+	}); err != nil {
+		return errors.Wrap(err, "get pod list")
+	}
+
+	if !stsChanged(currentSet, pods.Items) {
+		return nil
+	}
+
+	log.Info("statefulSet was changed, run smart update")
+
+	if cr.HAProxyEnabled() && cr.Status.HAProxy.State != apiv1alpha1.StateReady {
+		log.Info("Waiting for HAProxy to be ready before smart update")
+		return nil
+	}
+
+	if cr.RouterEnabled() && cr.Status.Router.State != apiv1alpha1.StateReady {
+		log.Info("Waiting for MySQL Router to be ready before smart update")
+		return nil
+	}
+
+	running, err := r.isBackupRunning(ctx, cr)
+	if err != nil {
+		log.Error(err, "can't start 'SmartUpdate'")
+		return nil
+	}
+	if running {
+		log.Info("can't start/continue 'SmartUpdate': backup is running")
+		return nil
+	}
+
+	if currentSet.Status.ReadyReplicas < currentSet.Status.Replicas {
+		log.Info("Can't start/continue 'SmartUpdate': waiting for all replicas are ready")
+		return nil
+	}
+
+	primaryHost, err := r.getPrimaryHost(ctx, cr)
+	if err != nil {
+		return err
+	}
+	idx, err := getPodIndexFromHostname(primaryHost)
+	if err != nil {
+		return err
+	}
+	primPod, err := getMySQLPod(ctx, r.Client, cr, idx)
+	if err != nil {
+		return errors.Wrap(err, "get primary pod")
+	}
+	log.Info("primary pod", "name", primPod.Name)
+
+	for _, pod := range pods.Items {
+		pod := pod
+		if pod.Name == primPod.Name {
+			continue
+		}
+
+		log.Info("apply changes to secondary pod", "pod name", pod.Name)
+
+		if pod.ObjectMeta.Labels[controllerRevisionHash] == sts.Status.UpdateRevision {
+			log.Info("pod updated updated", "pod name", pod.Name)
+			continue
+		}
+
+		return r.Client.Delete(ctx, &pod)
+	}
```

Comment:
this block needs to consider if pod is ready or not. what if we restarted a secondary pod, but it fails to become ready? we shouldn't delete another pod in this case.

Reply:
It is checked before, line 70. We will never come to this point if there is a non-ready pod.